### PR TITLE
Admin Guide for Recipes

### PR DESCRIPTION
**Pull request by jasonviviano**
## Description

- Overview
- Features
          - Input context parameter
          - Output values contract
          - Output resource contract
- Example
          - How to author a recipe template
          - How to publish a recipe template
          - How to register a recipe

### CODE REVIEW

File: docs/content/operations/recipes-creation/_index.md
Comment (shalabhms) [inline docs/content/operations/recipes-creation/_index.md]: Same comment as I mentioned on the Developer Guide PR. It would be good to state what problem Recipes is solving and then state how it it solving. 

So a reordering might help here

```suggestion
Recipes enables  **separation of concerns** between infrastructure teams and developers by allowing for a **automated infrastructure deployment** that doesn't require developers to pick up infrastructure resource expertise.
Recipe is a bicep template stored in container registry that can be deployed by [Links]({{< ref links-concept >}}) once a Recipe has been registered in a Radius environment. 
```

File: docs/content/operations/recipes-creation/_index.md
Comment (shalabhms) [inline docs/content/operations/recipes-creation/_index.md]: Please add the description of these properties inline here - `resource`, `application`, `environment`, `runtime`

File: docs/content/operations/recipes-creation/_index.md
Comment (shalabhms) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
title: "Radius Recipes"
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
linkTitle: "Author Recipes"
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
title: "Author Custom Radius Recipes"
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
description: "Learn how to author custom Recipe templates to automate infrastructure deployment"
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
```

This info is covered in the Recipes dev page, I'd remove this and instead add a reference below (see below comment)

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
Recipes enable a **separation of concerns** between infrastructure teams and developers by **automating infrastructure deployment**.

Learn more in the Recipes overview page:

{{< button page="recipes" text="Recipes overview" >}}
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
## Author a custom Recipe

This page will describe how to author and publish a custom Recipe template, letting you define how infrastructure should be deployed and configured within your environment.
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
### Select an Infrastructure as Code language

Today, Recipes only support the Bicep IaC language. Your custom Recipe template will begin with a new Bicep file, which will contain all of the resources you want to automatically deploy for each resource that runs the Recipe.

```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]: I think a full reference table would be better at the bottom of the page, instead of a smaller top-level table here.

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
### Ensure your resource names are **unique** and **repeatable**
Recipe will be called by multiple many resources and need to be generalized templates. The resources inside of a Recipe cannot interfere or overwrite each other. To make sure you don't accidently overwrite or corrupt your infrastructure, you will want to make sure **each resource in your Recipe contains unique names**.
Additionally, every time an application is deployed, Recipes are re-evaluated and deployed. Because IaC languages like Bicep are idempotent, if there are no changes to the template then nothing is re-deployed or re-configured. To make sure Recipes can be re-run, **each resource needs a repeatable name that is the same every time it is deployed**.
To make naming easy, a `context` parameter is automatically injected into your template.
It contains information about the backing resource which you can use to generate unique and repeatable names:
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
### Add parameters for Recipe customization (optional)

You can optionally choose to add parameters to your Recipe to allow developers or operators to specify additional configuration. Parameters can be set both when a Recipe is added to an environment by an operator, or by a developer when the Recipe is called.

You can create any [parameter type supported by Bicep](https://learn.microsoft.com/azure/azure-resource-manager/bicep/parameters):
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
@description('A custom parameter that can be set by a developer or operator')
param myParam string = 'default value'` 
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
### Output the target infrastructure

Once you've defined your backing infrastructure, you'll need to output it from your IaC template so it can be "wired-up" to the resource that called the Recipe.

Recipes allow you to either specify `values` or `resource`:

{{< tabs Values Resource >}}

{{% codetab %}}
When you output a `values` object, all of the individual properties will be directly mapped to the resource calling the Recipe. This allows you that most control over the resource being created within the application.

Simply define an object that matches the schema of the resource calling the Recipe. For example, for an `Application.Link/redisCaches` resource, a Recipe would output:
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]: Should these be nested under a `secrets` property?

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
Some Radius resources, such as Links, may support automatically mapping infrastructure via a resource ID. Refer to the backing resource's documentation to know what infrastructure is supported.

For supported infrastructure and resources, you can simply output the resource ID as a string, and the properties will automatically configured. For example, `Applications.Link/redisCaches` supports auto-mapping of `Microsoft.Cache/redis`:

```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
resource azureCache `Microsoft.Cache/redis@2022-06-01' = {...}

output resource string = azureCache.id
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
If you specify both `values` and `resource`, `resource` will be used first and then `values` will override.
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
```

I'd remove this as it's now redundant

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
### Store your template in a Bicep registry
```

I think for now let's use the exsiting "Bicep registry" term that's already well-understood for Bicep authors and in the Bicep documentation.

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
Recipes leverage [Bicep registries](https://learn.microsoft.com/azure/azure-resource-manager/bicep/private-module-registry) for template storage. Once you've authored a Recipe, you can publish it to your preferred OCI-compliant registry.

For private registries, make sure the cloud provider configured in your Radius environment has pull permissions for your container registry.

You can use the `az bicep publish` to easily publish your Recipe Bicep template:
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
az bicep publish --file myrecipe.bicep --target myregistry.azurecr.io/recipes/myrecipe:v1
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
```bash
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
### Register your Recipe with your environment
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion

Now that your Recipe Bicep template has been stored within your Bicep registry, you can add it your Radius environment to be used by developers. This allows you to mix-and-match templates for each of your environments such as dev, canary, and prod.

Recipes can be added via the rad CLI, or via an environment Bicep definition:

```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
```bash
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]: Also please add a Bicep environment template example here too.

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion

- Visit the Recipe developer guide for more information.
```

This should also be a `{{< ref >}} to this page (make sure this PR is based off of the dev PR branch, or wait until that is merged and then pull in the latest changes from `edge`)

**Review by AaronCrawfis — CHANGES_REQUESTED** [review]
Great start! Added some notes on how to streamline this

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
### Ensure your resource names are **unique** and **repeatable**

```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
Recipes need to be generalized templates, as multiple resources can run each Recipe. The resources inside of a Recipe cannot interfere or overwrite each other. To make sure you don't accidentally overwrite or corrupt your infrastructure, you will want to make sure **each resource in your Recipe contains unique names**.

```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
Additionally, every time an application is deployed, Recipes are re-evaluated and re-deployed. Because IaC languages like Bicep are idempotent, if there are no changes to the template, then nothing is changed. To make sure Recipes can be re-run, **each resource needs a repeatable name that is the same every time it is deployed**.

```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
To make naming easy, a `context` parameter is automatically injected into your template. It contains information about the backing resource which you can use to generate unique and repeatable names.

For example, you can use `context.resource.id` to generate a unique and repeatable name:

{{< rad file="snippets/=template.bicep" embed="true" marker="//NAME" >}}
```

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
```

**Review by AaronCrawfis — CHANGES_REQUESTED** [review]

File: docs/content/operations/recipes-creation/_index.md
Comment (kachawla) [reply docs/content/operations/recipes-creation/_index.md]: Yes. And host and port nested under `values`. https://github.com/project-radius/radius/blob/main/test/functional/corerp/resources/testdata/recipes/test-recipes/redis-recipe-value-backed.bicep#L61-L73

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion
```

Be careful not to have repeated blank lines

File: docs/content/operations/recipes-creation/_index.md
Comment (AaronCrawfis) [inline docs/content/operations/recipes-creation/_index.md]:
```suggestion

```

**Review by AaronCrawfis — APPROVED** [review]

**Pull request merged**
